### PR TITLE
feat: recreate flutter create with sub commands

### LINKED ISSUE #490: feat(create)!: use subcommands for different template types & Proxy
## Proposal

By splitting the `create` command to use subcommands we can improve the usage and options to be subcommand-specific. 

Since there are some common options for all `create` commands (`<project name> --output-directory`) and lots of options that are template-specific (platforms for the plugin, executable name for the CLI, etc.) this approach will allow us to organize the command in a way that better aligns with the intended usage/structure.

### Proposed Usage

```sh
# Top Level Create Usage
very_good create <subcommand> <project-name> <options>

options:
  --verbose (-v)
  --help (-h)
  --output-directory (-o)
```

```sh
# Very Good Core Usage
very_good create flutter_app <project-name> <options>

options:
    --template="core"
    --org-name
    --description
```

```sh
# Very Good Dart Package Usage
very_good create dart_pkg <project-name> <options>

options:
    --description
```

```sh
# Very Good Dart CLI Usage
very_good create dart_cli <project-name> <options>

options:
    --executable-name
    --description
```

```sh
# Very Good Flutter Package Usage
very_good create flutter_pkg <project-name> <options>

options:
    --description
```

```sh
# Very Good Flutter Plugin Usage
very_good create flutter_plugin <project-name> <options>

options:
    --org-name
    --description
    --platforms
```

```sh
# Very Good Docs Site Usage
very_good create docs_site <project-name> <options>

options:
    --org-name
    --description
```

```sh
# Very Good Flame Game Usage
very_good create flame_game <project-name> <options>

options:
    --org-name
    --description
```

- [ ]  Proxy all the old create calls to the new subcommands.

### LINKED ISSUE #582: feat(create)! subcommand: Very Good Core Usage
Split the create command to use subcommands.

```sh
# Very Good Core Usage
very_good create flutter_app <project-name> <options>

options:
    --template="core"
    --org-name
    --description
```

### PR DESCRIPTION
**Pull request by renancaraujo**
## Description

feat: refactor flutter create with subcommands

Add a common superclass for create subcommands, preserve legacy behavior with a deprecated message

Closes #490
Closes #582
Closes [#583](https://github.com/VeryGoodOpenSource/very_good_cli/issues/583)

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [x] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: lib/src/commands/create/create_subcommand.dart [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    return result;
+  }
+
+  Future<int> runCreate(MasonGenerator generator, Template template) async {
+    final generateProgress = logger.progress('Bootstrapping');
+    final outputDirectory = this.outputDirectory;
+
+    var vars = getTemplateVars();
+
+    final target = DirectoryGeneratorTarget(outputDirectory);
```
Comment (felangel) [inline lib/src/commands/create/create_subcommand.dart:181]:
```suggestion
    var vars = getTemplateVars();

    final target = DirectoryGeneratorTarget(this.outputDirectory);
```

File: lib/src/commands/create/commands/flutter_app.dart [hunk 2]
```diff
@@ -0,0 +1,50 @@
+// ignore_for_file: public_member_api_docs
+
+import 'package:mason/mason.dart';
+import 'package:usage/usage.dart';
+import 'package:very_good_cli/src/commands/create/create_subcommand.dart';
+import 'package:very_good_cli/src/commands/create/templates/templates.dart';
+
+class CreateFlutterApp extends CreateSubCommand with OrgName {
```
Comment (felangel) [inline lib/src/commands/create/commands/flutter_app.dart:8]: I love how this turned out! Looks so much cleaner and easier to extend 💯

File: lib/src/command_runner.dart [hunk 3]
```diff
@@ -5,6 +5,7 @@ import 'package:mason/mason.dart' hide packageVersion;
 import 'package:pub_updater/pub_updater.dart';
 import 'package:usage/usage_io.dart';
 import 'package:very_good_cli/src/commands/commands.dart';
+import 'package:very_good_cli/src/commands/create/create_legacy.dart';
```
Comment (felangel) [inline lib/src/command_runner.dart:8]: do we need this additional import or can it be added to the commands.dart barrel?

File: lib/src/commands/create/create.dart [hunk 4]
```diff
@@ -1,314 +1,57 @@
...
-final RegExp _identifierRegExp = RegExp('[a-z_][a-z0-9_]*');
-final RegExp _orgNameRegExp = RegExp(r'^[a-zA-Z][\w-]*(\.[a-zA-Z][\w-]*)+$');
-
-/// A method which returns a [Future<MasonGenerator>] given a [MasonBundle].
-typedef MasonGeneratorFromBundle = Future<MasonGenerator> Function(MasonBundle);
-
-/// A method which returns a [Future<MasonGenerator>] given a [Brick].
-typedef MasonGeneratorFromBrick = Future<MasonGenerator> Function(Brick);
+import 'package:very_good_cli/src/commands/create/commands/commands.dart';
+import 'package:very_good_cli/src/commands/create/create_subcommand.dart';
```
Comment (felangel) [inline lib/src/commands/create/create.dart:5]: do we need this import or can it be added to the commands.dart barrel?

File: lib/src/commands/create/create.dart [hunk 5]
```diff
@@ -1,314 +1,57 @@
...
-      ..addFlag(
-        'publishable',
-        negatable: false,
-        help: 'Whether the generated project is intended to be published '
-            '(Does not affect flutter application templates)',
-      );
+  }) {
+    // Legacy sub command: hidden sub command that maintains backwards
+    // compatibility with the `very_good create <project-name>` command syntax.
+    // The command runner will call legacy if the suer run the legacy syntax.
```
Comment (felangel) [inline lib/src/commands/create/create.dart:23]:
```suggestion
    // The command runner will call legacy if the user uses the legacy syntax.
```

**Review by felangel — DISMISSED** [review]
Left some minor comments but overall LGTM 🎉

File: lib/src/commands/create/create.dart (hunk 4 above)
Comment (renancaraujo) [reply lib/src/commands/create/create.dart:5]: Adding to the barrel

**Review by felangel — APPROVED** [review]

File: lib/src/commands/create/commands/create_subcommand.dart [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+
+    await template.onGenerateComplete(
+      logger,
+      Directory(path.join(target.dir.path, projectName)),
+    );
+
+    return ExitCode.success.code;
+  }
+
+  /// Responsible for returns the template parameters to be passed to the
```
Comment (erickzanardo) [inline lib/src/commands/create/commands/create_subcommand.dart:223]:
```suggestion
  /// Responsible for returning the template parameters to be passed to the
```

**Review by erickzanardo — APPROVED** [review]
One minor comment, but LGTM!

**Pull request merged**